### PR TITLE
Migrate xpallets-assets-registrar to pallet v2

### PR DESCRIPTION
**Pull request by VegeBun-csj**
fix some warnings

### CODE REVIEW

File: xpallets/assets-registrar/src/lib.rs [hunk 1]
```diff
@@ -300,17 +332,30 @@ impl<T: Config> Module<T> {
...
+    /// Direct implementation of `GenesisBuild::assimilate_storage`.
+    ///
+    /// Kept in order not to break dependency.
+    pub fn assimilate_storage<T: Config>(
+        &self,
+        storage: &mut sp_runtime::Storage,
+    ) -> Result<(), String> {
+        <Self as GenesisBuild<T>>::assimilate_storage(self, storage)
+    }
+}
```
Comment (hacpy) [inline xpallets/assets-registrar/src/lib.rs:363]: According to the suggestions of the Rust compiler, you may be able to replace the re-implementation ``assimilate_storage`` by adding use ``crate::mock::sp_api_hidden_includes_construct_runtime::hidden_include::traits::GenesisBuild;`` in mock.rs.

File: xpallets/assets/src/lib.rs [hunk 2]
```diff
@@ -271,18 +271,18 @@ impl<T: Config> Module<T> {
...
+            .filter_map(|(id, _info)| {
                 if id == T::NativeAssetId::get() {
                     // ignore native asset
                     None
                 } else {
                     let data = (
                         id,
                         TotalAssetInfo {
-                            info,
+                            info: xpallet_assets_registrar::AssetInfo::default(),
```
Comment (hacpy) [inline xpallets/assets/src/lib.rs]: If there is no other reason, please do not modify the value arbitrarily, it is very dangerous.

**Review by hacpy — COMMENTED** [review]
If the modification is over, finally check again according to [upgrade-guidelines](https://crates.parity.io/frame_support/attr.pallet.html#checking-upgrade-guidelines).

File: xpallets/assets-registrar/src/lib.rs (hunk 1 above)
Comment (hacpy) [reply xpallets/assets-registrar/src/lib.rs:363]: Better to use ``frame_support::traits::GenesisBuild;``

File: xpallets/assets-registrar/src/lib.rs [hunk 3]
```diff
@@ -19,118 +19,59 @@ pub mod weights;
...
-}
+    #[pallet::pallet]
+    #[pallet::generate_store(pub(super) trait Store)]
+    pub struct Pallet<T>(_);
 
-decl_storage! {
-    trait Store for Module<T: Config> as XAssetsRegistrar {
-        /// Asset id list for each Chain.
-        pub AssetIdsOf get(fn asset_ids_of): map hasher(twox_64_concat) Chain => Vec<AssetId>;
+    #[pallet::config]
```
Comment (hacpy) [inline xpallets/assets-registrar/src/lib.rs:48]:
```suggestion
    /// The module's config trait.
    ///    
    /// `frame_system::Trait` should always be included in our implied traits.
    #[pallet::config]
```

File: xpallets/assets-registrar/src/lib.rs [hunk 4]
```diff
@@ -219,13 +162,102 @@ decl_module! {
...
+    pub enum Event<T: Config> {
+        /// A new asset was registered. [asset_id, has_mining_rights]
+        Registered(AssetId, bool),
+        /// A deregistered asset was recovered. [asset_id, has_mining_rights]
+        Recovered(AssetId, bool),
+        /// An asset was deregistered. [asset_id]
+        Deregistered(AssetId),
+    }
+
+    #[pallet::error]
```
Comment (hacpy) [inline xpallets/assets-registrar/src/lib.rs:183]:
```suggestion
    /// Error for the XAssetRegistrar Module
    #[pallet::error]
```

File: xpallets/assets-registrar/src/lib.rs [hunk 5]
```diff
@@ -19,118 +19,59 @@ pub mod weights;
...
-        AssetDoesNotExist,
-        /// The asset is already valid (online), no need to recover.
-        AssetAlreadyValid,
-        /// The asset is invalid (not online).
-        AssetIsInvalid,
-    }
-}
+    #[pallet::pallet]
+    #[pallet::generate_store(pub(super) trait Store)]
+    pub struct Pallet<T>(_);
```
Comment (0xrjman) [inline xpallets/assets-registrar/src/lib.rs]: `pub struct Pallet<T>(_);` won't make your compilation fail, but it's better to use `pub struct Pallet<T>(PhantomData<T>);` instead. `PhantomData` is there to let you tell the compiler information that it cannot infer itself, correctly expressing lifetimes in `PhantomData` is a part of creating a safe abstraction around unsafe code.

File: xpallets/assets/src/lib.rs (hunk 2 above)
Comment (0xrjman) [reply xpallets/assets/src/lib.rs]: this is a problem

File: xpallets/assets-registrar/src/lib.rs (hunk 5 above)
Comment (0xrjman) [inline xpallets/assets-registrar/src/lib.rs]:
```suggestion
    pub struct Pallet<T>(PhantomData<T>);
```

File: xpallets/assets/src/lib.rs (hunk 2 above)
Comment (VegeBun-csj) [reply xpallets/assets/src/lib.rs:283]: because it need a `AssetInfo` type

File: xpallets/assets-registrar/src/lib.rs [hunk 6]
```diff
@@ -219,13 +162,102 @@ decl_module! {
...
+        /// A new asset was registered. [asset_id, has_mining_rights]
+        Registered(AssetId, bool),
+        /// A deregistered asset was recovered. [asset_id, has_mining_rights]
+        Recovered(AssetId, bool),
+        /// An asset was deregistered. [asset_id]
+        Deregistered(AssetId),
+    }
+
+    #[pallet::error]
+    /// Error for the XAssetRegistrar Module
```
Comment (0xrjman) [inline xpallets/assets-registrar/src/lib.rs]:
```suggestion
    /// Error for the XAssetRegistrar Pallet
```

File: xpallets/assets-registrar/src/lib.rs [hunk 7]
```diff
@@ -219,13 +162,102 @@ decl_module! {
...
+            for (id, asset, is_online, has_mining_rights) in &self.assets {
+                Pallet::<T>::register(
+                    frame_system::RawOrigin::Root.into(),
+                    *id,
+                    asset.clone(),
+                    *is_online,
+                    *has_mining_rights,
+                )
+                .expect("asset registeration during the genesis can not fail");
+            }
```
Comment (0xrjman) [inline xpallets/assets-registrar/src/lib.rs]: There is no problem logically, but it is best to follow the implementation of the template

File: runtime/chainx/src/lib.rs [hunk 8]
```diff
@@ -1076,7 +1076,7 @@ construct_runtime!(
 
         // ChainX basics.
         XSystem: xpallet_system::{Pallet, Call, Storage, Event<T>, Config} = 24,
-        XAssetsRegistrar: xpallet_assets_registrar::{Pallet, Call, Storage, Event, Config} = 25,
+        XAssetsRegistrar: xpallet_assets_registrar::{Pallet, Call, Storage, Event<T>, Config} = 25,
```
Comment (rjman-ljm) [inline runtime/chainx/src/lib.rs:1079]: `type Event: From<Event<Self>> + IsType<<Self as frame_system::Config>::Event>;` has defined in pallet::Config, it's unnecessory

**Pull request closed without merging**